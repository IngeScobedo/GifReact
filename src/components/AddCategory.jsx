/* eslint-disable react/no-unknown-property */
import { useState } from 'react'
import PropTypes from 'prop-types'

export const AddCategory = ({ onNewCategory }) => {
  const [inputValue, setInputValue] = useState('')

  const handleInputChange = ({ target }) => {
    console.log(target.value)
    setInputValue(target.value)
  }

  const handleSubmit = (event) => {
    event.preventDefault()
    const value = inputValue.trim()
    console.log('value', value)
    if (value.length <= 1) { return }
    onNewCategory(value)
    setInputValue('')
  }

  return (
    <form aria-label="form" onSubmit={handleSubmit} className="w-full flex justify-center">
        <input
            type="text"
            id="search"
            name="search"
            className='border border-black rounded-[28px] text-center text-black my-4 mx-2 w-[80%] px-2 py-3'
            placeholder='Buscar gifs...'
            value={inputValue}
            onChange={ handleInputChange }
            />
    </form>
  )
}

AddCategory.propTypes = {
  onNewCategory: PropTypes.func.isRequired
}
