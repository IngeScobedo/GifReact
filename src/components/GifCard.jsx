import PropTypes from 'prop-types'

export const GifCard = ({ url, title }) => {
  return (
    <div className="w-full h-[145px] rounded-md border border-slate-200 flex flex-col">
        <div className="h-[70%] w-full rounded-t-md">
            <img src={url} alt={title} className='w-full h-full object-cover rounded-t-md' />
        </div>
        <div className='w-full h-[30%] flex flex-col items-center px-2 py-1 text-center justify-center bg-white rounded-b-md'>
            <h2 className='line-clamp-1'>{title}</h2>
        </div>
    </div>
  )
}

GifCard.propTypes = {
  url: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired
}
