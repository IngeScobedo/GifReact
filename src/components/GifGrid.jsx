import PropTypes from 'prop-types'
import { useFetchGifs } from '../hooks'
import { GifCard } from './GifCard'
import './GifGrid.css'

export const GifGrid = ({ category }) => {
  const { gifs, isLoading } = useFetchGifs(category)

  return (
    <div className="w-full flex justify-center bg-gray-200">
        <section className="w-full max-w-[1100px] flex flex-col gap-5 px-8 py-3">
            <div className='w-full flex justify-center text-center'>
                <h1
                    className='text-indigo-500 font-bold text-4xl first-letter:uppercase' >{category}
                </h1>
            </div>
            <div className='w-full grid grid-cols-3 gap-4'>
                {
                    !isLoading
                      ? gifs.map(gif => (<GifCard url={gif.url} title={gif.title} key={gif.id}/>))
                      : (
                        <div className='w-full h-20 flex justify-center items-center'>
                            <span className='loader'></span>
                        </div>
                        )
                }
            </div>
        </section>

    </div>
  )
}

GifGrid.propTypes = {
  category: PropTypes.string.isRequired
}
