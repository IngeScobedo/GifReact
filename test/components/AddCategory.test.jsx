import { fireEvent, render, screen } from '@testing-library/react'
import { AddCategory } from '../../src/components/AddCategory'

describe('Pruebas en <AddCategory/>', () => {
  const query = 'Hunter x Hunter'

  test('Debe hace match con el snapshot', () => {
    const onNewCategory = jest.fn()
    const { container } = render(<AddCategory onNewCategory={onNewCategory}/>)
    expect(container).toMatchSnapshot()
  })

  test('Debe de cambiar el valor de la caja de texto al escribir', () => {
    const onNewCategory = jest.fn()
    render(<AddCategory onNewCategory={onNewCategory}/>)
    const input = screen.getByRole('textbox')
    fireEvent.input(input, { target: { value: query } })
    expect(input.value).toBe(query)
  })

  test('Debe de hacer el submit del formulario si la caja de texto tiene mas de 1 carácter', () => {
    const onNewCategory = jest.fn()
    render(<AddCategory onNewCategory={onNewCategory}/>)

    const input = screen.getByRole('textbox')
    const form = screen.getByRole('form')

    fireEvent.input(input, { target: { value: query } })
    fireEvent.submit(form)

    expect(input.value).toBe('')
    expect(onNewCategory).toHaveBeenCalled()
    expect(onNewCategory).toHaveBeenCalledTimes(1)
    expect(onNewCategory).toHaveBeenCalledWith(query)
  })

  test('El onNewCategory no debe ser llamado si la caja de texto esta vació', () => {
    const onNewCategory = jest.fn()
    render(<AddCategory onNewCategory={onNewCategory}/>)

    const input = screen.getByRole('textbox')
    const form = screen.getByRole('form')

    expect(input.value).toBe('')

    fireEvent.submit(form)

    expect(onNewCategory).toHaveBeenCalledTimes(0)
  })
})
