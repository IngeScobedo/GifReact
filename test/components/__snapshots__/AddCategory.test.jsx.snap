// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <AddCategory/> Debe hace match con el snapshot 1`] = `
<div>
  <form
    aria-label="form"
    class="w-full flex justify-center"
  >
    <input
      class="border border-black rounded-[28px] text-center text-black my-4 mx-2 w-[80%] px-2 py-3"
      id="search"
      name="search"
      placeholder="Buscar gifs..."
      type="text"
      value=""
    />
  </form>
</div>
`;
